src/container/srv_container.c

```diff
@@ -3628,6 +3659,294 @@ ds_cont_list(uuid_t pool_uuid, struct daos_pool_cont_info **conts,
+			whole_match ? "matched" : "did not match");
+		*match = whole_match;
+	}
+
+	return rc;
+}
+
+/* argument type for callback function to filter containers */
+struct filter_cont_iter_args {
+	uuid_t				 pool_uuid;
+
+	/* Filter criteria specification */
+	daos_pool_cont_filter_t		*filt;
+
+	/* Number of containers in pool and conts[] index while counting */
+	uint64_t			 ncont;
+
+	/* conts[]: capacity*/
+	uint64_t			 conts_len;
+	struct daos_pool_cont_info2	*conts;
+	struct cont_svc			*svc;
+	struct rdb_tx			*tx;
+};
+
+/* callback function for filter containers iteration. */
+static int
+filter_cont_cb(daos_handle_t ih, d_iov_t *key, d_iov_t *val, void *varg)
+{
+	struct filter_cont_iter_args	*ap = varg;
+	struct daos_pool_cont_info2	*pcinfo;
+	uuid_t				 cont_uuid;
+	struct cont			*cont;
+	daos_prop_t			*prop = NULL;
+	bool				 filt_match = false;
+	int				 rc;
+	(void)val;
+
+	if (key->iov_len != sizeof(uuid_t)) {
+		D_ERROR("invalid key size: key="DF_U64"\n", key->iov_len);
+		return -DER_IO;
+	}
+	uuid_copy(cont_uuid, key->iov_buf);
+
+	D_DEBUG(DB_MD, "pool/cont: "DF_CONTF": ncont=%zu, conts_len=%zu\n",
+		DP_CONT(ap->pool_uuid, cont_uuid), ap->ncont, ap->conts_len);
+
+	/* Lookup container, see if it matches filter specification before adding to ap->conts[] */
+	rc = cont_lookup(ap->tx, ap->svc, cont_uuid, &cont);
+	if (rc != 0) {
+		D_ERROR(DF_CONT": lookup cont failed, "DF_RC"\n",
+			DP_CONT(ap->pool_uuid, cont_uuid), DP_RC(rc));
+		return rc;
+	}
+
+	rc = cont_filter_match(ap->tx, cont, ap->filt, &filt_match);
+	if (rc != 0) {
+		D_ERROR(DF_CONT": cont_match() failed, "DF_RC"\n",
+			DP_CONT(ap->pool_uuid, cont_uuid), DP_RC(rc));
+		goto out_cont;
+	} else if (filt_match == false) {
+		goto out_cont;
+	}
+
+	/* The container matches the filter specification: save information */
+
+	/* Realloc conts[] if needed (double each time starting with 1) */
+	if (ap->ncont == ap->conts_len) {
+		void	*ptr;
+		size_t	realloc_elems = (ap->conts_len == 0) ? 1 : ap->conts_len * 2;
+
+		D_REALLOC_ARRAY(ptr, ap->conts, ap->conts_len, realloc_elems);
+		if (ptr == NULL)
+			D_GOTO(out_cont, rc = -DER_NOMEM);
+
+		ap->conts = ptr;
+		ap->conts_len = realloc_elems;
+	}
+
+	pcinfo = &ap->conts[ap->ncont];
+	ap->ncont++;
+	uuid_copy(pcinfo->pci_id.pci_uuid, cont_uuid);
+
+	/* TODO: Specify client cont_proto_version. This is invoked from a pool client RPC */
+	rc = cont_info_read(ap->tx, cont, DAOS_CONT_VERSION /* engine protocol version */,
+			    &pcinfo->pci_cinfo);
+	if (rc != 0) {
+		D_ERROR(DF_CONT": read container info failed, "DF_RC"\n",
+			DP_CONT(ap->pool_uuid, cont_uuid), DP_RC(rc));
+		goto out_cont;
+	}
+
+	rc = cont_prop_read(ap->tx, cont, DAOS_CO_QUERY_PROP_LABEL, &prop, true);
+	if (rc != 0) {
+		D_ERROR(DF_CONT": cont_prop_read() failed, "DF_RC"\n",
+			DP_CONT(ap->pool_uuid, cont_uuid), DP_RC(rc));
+		goto out_cont;
+	}
+
+	strncpy(pcinfo->pci_id.pci_label, prop->dpp_entries[0].dpe_str, DAOS_PROP_LABEL_MAX_LEN);
```

Comment:
Right now,  “container_label_not_set” is set as the label when a user does not provide label when creates a container. I have a PR that will allows container label to be empty, so "prop->dpp_entries[0].dpe_str" could be NULL when a container has label not set.

Reply:
testing for NULL dpe_str in the latest version of the patch